---
title: "Week 10 Questions"

format:
  html:
    code-fold: false
    code-tools: true
    code-copy: true
    highlight-style: github
    code-overflow: wrap
---

1.  **Looking back on the quarter what are two things you were not expecting to learn but did, glad you did, and how would you implement this knowledge beyond this class?**

I wasn't expecting to learn so much about preparing and publishing workflows (via quarto documents, presentations, and **rpubs** in particular). Mark covered the research compendium stuff last quarter in his environmental data class, and both of these topics together will allow me to document and share my compendium and workflow with ease.

2.  **Rank the following from 1-5, with 1 being your best friend and 5 being your enemy (raven, rstudio, knitr, bash, github)**

(1) --> (5): rstudio, github, raven, knitr (lol knitting still perplexes me), bash (but bash still feels like the ultimate frenemy)

3.  **As part of your research project you must of learned something beyond methods. What is the most important biological finding from your work to to date?** Assuming you follow this line of research what is your long term goal of what you expect to learn about biology, ecology, or the environment?

With the help of Bryan (this is his data!) we actually noticed that by entirely removing one site that was being problematic in the alignment and sorting phase (East Sound), the portfolio of retained biallelic SNPs seemed to change significantly. I retained *more biallelic SNPs* across the other 15 sites than Bryan was able to because, we think, that East Sound had such unique variation that lots of SNPs were characterized as triallelic. We aren't interested in triallelic SNPs because most modeling approaches in pop gen rely on two alleles p and q. 

I am really interested in understanding adaptive variation in populations, and leveraging that knowledge to design restoration and conservation efforts that promote *resilience* in an era of climate change. 
