---
title: "week07 - motifs"
format: html
editor: visual
---

```{r, engine = "bash"}

cd ../data
curl -O https://gannet.fish.washington.edu/panopea/Cg-roslin/cgigas_uk_roslin_v1_genomic-mito.fa

```
```{bash}
pwd
head data/cgigas_uk_roslin_v1_genomic-mito.fa
```

```{r}
library(seqinr)

# Replace 'input.fasta' with the name of your multi-sequence fasta file
input_file <- "data/cgigas_uk_roslin_v1_genomic-mito.fa"
sequences <- read.fasta(input_file)

```

```{r}
# Set the seed for reproducibility (optional)
set.seed(42)

number_of_sequences_to_select <- 10

if (length(sequences) < number_of_sequences_to_select) {
  warning("There are fewer than 10 sequences in the fasta file. All sequences will be selected.")
  number_of_sequences_to_select <- length(sequences)
}

selected_indices <- sample(length(sequences), number_of_sequences_to_select)
selected_sequences <- sequences[selected_indices]

```

```{r}
# Replace 'output.fasta' with your desired output file name
output_file <- "output/10-seq-output.fasta"
write.fasta(selected_sequences, names(selected_sequences), output_file, open = "w")
```

### create an index for your 10-sequence fasta
#### index necessary for rendering in IGV. Not necessary for finding motifs.

```{bash}
#needed downstream for IGV
/home/shared/samtools-1.12/samtools faidx \
output/10-seq-output.fasta
```

### program that we are using to find the motifs.

```{bash}
fuzznuc -sequence output/10-seq-output.fasta -pattern CG -rformat gff -outfile output/10-seq-output.gff
```
This output file (.gff) provides the location for each cg motif that was found.

